clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/task/datawriter/CompositeVeniceWriter.java

```diff
@@ -1,41 +1,66 @@
+import static com.linkedin.venice.writer.VeniceWriter.DEFAULT_UPSTREAM_KAFKA_CLUSTER_ID;
+import static com.linkedin.venice.writer.VeniceWriter.DEFAULT_UPSTREAM_OFFSET;
+
 import com.linkedin.venice.annotation.NotThreadsafe;
 import com.linkedin.venice.exceptions.VeniceException;
 import com.linkedin.venice.pubsub.api.PubSubProduceResult;
 import com.linkedin.venice.pubsub.api.PubSubProducerCallback;
+import com.linkedin.venice.utils.lazy.Lazy;
+import com.linkedin.venice.views.VeniceView;
 import com.linkedin.venice.writer.AbstractVeniceWriter;
+import com.linkedin.venice.writer.ComplexVeniceWriter;
 import com.linkedin.venice.writer.DeleteMetadata;
+import com.linkedin.venice.writer.LeaderMetadataWrapper;
 import com.linkedin.venice.writer.PutMetadata;
 import com.linkedin.venice.writer.VeniceWriter;
 import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Future;
 import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.avro.generic.GenericRecord;
 
 
 /**
  * The composite writer contains a main writer and multiple child writers. The main writer will only perform the write
- * once all of its child writers are complete.
- * TODO The child writers are view writers. Ideally to avoid code duplication we should be using an array of
- * VeniceViewWriter here. However, the current implementation of VeniceViewWriter involves PCS which is something
- * specific to the ingestion path that we don't want to leak into venice-common.
+ * once all of its child writers are complete. Child writers are {@link com.linkedin.venice.writer.ComplexVeniceWriter}.
+ * This is to provide chunking support during NR pass-through. All records produced by the {@link ComplexVeniceWriter}'s
+ * main writer is expected to carry the
+ * {@link com.linkedin.venice.pubsub.api.PubSubMessageHeaders#VENICE_VIEW_PARTITIONS_MAP_HEADER}.
  */
 @NotThreadsafe
 public class CompositeVeniceWriter<K, V, U> extends AbstractVeniceWriter<K, V, U> {
   private final VeniceWriter<K, V, U> mainWriter;
-  private final AbstractVeniceWriter<K, V, U>[] childWriters;
+  private final ComplexVeniceWriter<K, V, U>[] childWriters;
   private final PubSubProducerCallback childCallback;
+  private final BiFunction<V, Integer, GenericRecord> deserializeFunction;
+  private final Function<V, V> decompressFunction;
 
   public CompositeVeniceWriter(
       String topicName,
       VeniceWriter<K, V, U> mainWriter,
-      AbstractVeniceWriter<K, V, U>[] childWriters,
-      PubSubProducerCallback childCallback) {
+      ComplexVeniceWriter<K, V, U>[] childWriters,
+      PubSubProducerCallback childCallback,
+      BiFunction<V, Integer, GenericRecord> deserializeFunction,
+      Function<V, V> decompressFunction) {
```

Comment:
Nit,
Can we consolidate these two params into one called something like `valueExtractor`?